fix(migrations): control flow migration omnibus bugfixes

jessicajaniuk opened the pull request:
This is an omnibus of minor formatting and bugfixes to the control flow migration:
1. preserves leading spaces on inline templates
2. Better handles self closing tags with or without the `/`
3. overall formatting fixes
4. adds support for `*ngFor="x of y; template: templateName"`
5. fixes a bug with building conditions when trailing semicolons are present for NgIf with else and then else
6. handles cases of transplanted views
7. Fixes loss of let variable on NgIf else and if then else
8. Handles case of ngif else or if then else not using semicolon separators
9. Ensures common module is removed from classes using templateUrl

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

thePunderWoman committed:
fix(migrations): control flow migration formatting fixes

This fix preserves leading indents in inline templates and also adds better handling for self closing tags

thePunderWoman committed:
fix(migrations): Add ngForTemplate support to control flow migration

This adds code to cover the rare use case of an ngFor with a template param.
fixes: #53068

JoostK commented on packages/core/schematics/test/control_flow_migration_spec.ts:
```diff
@@ -1703,6 +1703,45 @@ describe('control flow migration', () => {
+
+      const expected = [
+        `@for (item of items; track item) {`,
+        `  <ng-container *ngTemplateOutlet="itemTemplate; context: {$implicit: item}">`,
```
Might be a bit tricky, but `NgTemplateOutlet` or `CommonModule` needs to be imported for this to work.

jessicajaniuk replied:
Agree, though I think this use case is rare enough that if they have to manually add that in, it should be fine for the purposes of the migration.

pweyrich commented on packages/core/schematics/test/control_flow_migration_spec.ts:
```diff
@@ -1635,6 +1703,45 @@ describe('control flow migration', () => {
+      const actual = tree.readContent('/comp.html');
+
+      const expected = [
+        `@for (item of items; track item) {`,
+        `  <ng-container *ngTemplateOutlet="itemTemplate; context: {$implicit: item}">`,
+        `    <p>Stuff</p>`,
+        `  </ng-container>`,
+        `}\n`,
```
It should actually migrate to:
```suggestion
        `@for (item of items; track item) {`,
        `  <ng-container *ngTemplateOutlet="itemTemplate; context: {$implicit: item}">`,
        `  </ng-container>`,
        `}\n`,
        `<ng-template #itemTemplate let-item>`,
        `  <p>Stuff</p>`,
        `</ng-template>`
```

However, in case the template ref variable "itemTemplate" is apparently only used by that single ngFor, it could be simplified to:
```suggestion
        `@for (item of items; track item) {`,
        `  <p>Stuff</p>`,
        `}\n`,
```
But you really would have to make sure it's not used anywhere else to apply this simplification. Otherwise it would need to be retained.

jessicajaniuk replied:
Thanks for this! I've updated the migration to now match this behavior.

thePunderWoman committed:
fix(migrations): fixes control flow migration common module removal

Common module removal would not happen when a component used a templateUrl due to the checks being in separate files. This change passes the removal analysis back to the original source file to safely remove CommonModule.

thePunderWoman committed:
fix(migrations): allows colons in ngIf else cases to migrate

This makes sure colons after else and then cases get migrated properly.
fixes: #53150

AndrewKushnir commented on packages/core/schematics/ng-generate/control-flow-migration/types.ts:
```diff
@@ -114,18 +114,39 @@ export class ElementToMigrate {
+    let condition = chunks[0];
+
+    // checks for case of no usage of `;` in if else / if then else
+    const elseIx = condition.indexOf(' else ');
+    const thenIx = condition.indexOf(' then ');
```
I'm wondering if we should make leading whitespaces optional to capture scenarios like this: `;else ...` (similar to what you have on line 130).

jessicajaniuk replied:
These cases are specifically to detect the situation when they've omitted the `;`, and in that case, if there was no space, else and then would be part of the prior symbol. eg `*ngIf="blahelse stuff`.

crisbeto commented on packages/core/schematics/ng-generate/control-flow-migration/index.ts:
```diff
@@ -113,6 +118,12 @@ function runControlFlowMigration(
+function sortFilePaths(names: string[]): string[] {
+  const templateFiles = names.filter(n => n.endsWith('.html'));
+  const classFiles = names.filter(n => !n.endsWith('.html'));
+  return [...templateFiles, ...classFiles];
```
We should be able to sort these in place like `names.sort((a, b) => a.endsWith('.html') ? -1 : 0)`. Not a blocker.

jessicajaniuk replied:
Updated! Thank you.

jessicajaniuk commented:
Caretaker Note: This requires a [patch CL](https://critique.corp.google.com/cl/585691750) to land when this syncs in.

AndrewKushnir commented on packages/core/schematics/ng-generate/control-flow-migration/types.ts:
```diff
@@ -114,18 +114,39 @@ export class ElementToMigrate {
     if (secondStr) {
       const secondTargetLocation = this.attr.value.indexOf(secondStr);
-      return this.attr.value.slice(targetLocation + targetStr.length, secondTargetLocation).trim();
+      return this.attr.value.slice(targetLocation + targetStr.length, secondTargetLocation)
+          .replace(':', '')
+          .trim()
+          .split(';')[0]
+          .trim();
```
nit: we can probably add a small util function to avoid duplication with the code below.

jessicajaniuk replied:
I simplified this code instead so that there's no need to duplicate.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 53912fdf74e1224f95edc3dd550c51451e613c44.
